Increase the timeout value for service start/stop

lipingxue opened the pull request:
Fixes #1766 
This PR includes: 
Increase the timeout value when waiting for Samba service to start/stop and also the timeout value for block waiting.

The Swarm samba service can take up to 1.5 minutes to come up on setup with less resource, so set the timeout value for waiting for Samba service to start/stop to 90 seconds, and set the timeout value for block waiting for 120 seconds.

lipingxue commented:
@msterin This will happen one the first mount to a volume created by shared plugin.  The first mount request need to start a swarm service to run Samba server,  and the mount need to wait until samba server start successfully. In some setup,  it takes up to 90 seconds for the Samba server to come up. With a smaller timeout value (40 seconds before this change),  the first mount failed since samba service cannot start before it times out.

lipingxue commented:
@msterin What do you mean set this as a plugin option? You mean user need to specify the value when installed the plugin?

msterin commented:
yes, plugin option is similar to "--config" in config.json for vDVS plugin. You can add "--service-start-timeout 120" there. 

on the "why" - what takes 60 sec ? Sounds WAY too slow. Is it because of creating a new volume the very first time, and the need to format it ? Can you check in ESX logs what takes time ?

shivanshu21 commented:
BTW this sloth is only observable on nimbus VMs which are resource starved. My local VMs are able to mount the volume for the first time in 9 seconds.

The subsequent mounts on the same node are almost instantaneous.

luomiao commented:
Even on the testbed I got from nimbus, it won't take that long :)
But previous when I tried another script which deploys testbed with very limited resource, it will hit the timeout.
So it's very platform dependent.

@lipingxue 
Can you add the change to userguide to add this new config flag information when you update the PR?

msterin commented:
> But previous when I tried another script which deploys testbed with very limited resource, it will hit the timeout. So it's very platform dependent.

And this exactly my point. We should (1) understand which steps is taking all that time when resource are low   (2) NOT set timeout for extreme cases. but instead make it reasonable and provide a config

lipingxue commented:
@msterin 
Hi Mark, I have added an option to config the timeout value when installing the plugin(I introduced an env variable VFILE_TIMEOUT_IN_SECOND for this):
```
root@sc-rdops-vm02-dhcp-52-237:~# docker plugin install --grant-all-permissions --alias vshared lipingxue/vsphere-shared:vshared-unmount1 VFILE_TIMEOUT_IN_SECOND=90
vshared-unmount1: Pulling from lipingxue/vsphere-shared
bb2a72c3caa5: Download complete 
Digest: sha256:21e762a406a6dc31202a93e52ae6893b61e1d9adeb515564859ba477bb1564cc
Status: Downloaded newer image for lipingxue/vsphere-shared:vshared-unmount1
Installed plugin lipingxue/vsphere-shared:vshared-unmount1
```
In the log, we can see the timeout value we used is the value from config variable
```
2017-08-23 11:08:51.919778216 -0700 PDT [INFO] Get sambaRequest timeout from config 
 value=90
```

lipingxue commented:
@msterin @luomiao I have addressed your comments, and please review it.

luomiao commented:
LGTM

msterin commented on client_plugin/drivers/shared/dockerops/dockerops.go:
```diff
@@ -65,14 +66,44 @@ const (
+		return SambaRequestTimeout
+	}
+	// read sambaRequest Timout value from enviroment variable
+	timeOutFromConfig := os.Getenv("VFILE_TIMEOUT_IN_SECOND")
```
I do not think there is a need for `isSambaRequestTimeoutSet `.
This should make do, won't it ? 
 
```go
func GetSambaRequestTimeout() time.Duration {
 timeOutSec, err := strconv.Atoi(os.Getenv("VFILE_TIMEOUT_IN_SECOND"))
 if err != nil {
   timeOutSec = defaultSambaRequestTimeOutInSecond 
 }
  log.WithFields(log.Fields{"value": timeOutSec ,}).Info("sambaRequest timeout")
  return time.Duration(timeOutSec ) * time.Second
}
```

  
Also, this hjas nothing to do with Samba, it is a generic timeout for service start. Please rename accordingly

msterin commented on client_plugin/drivers/shared/dockerops/dockerops.go:
```diff
@@ -65,14 +66,24 @@ const (
-	// Timeout to mark Samba service launch as unsuccessful
-	sambaRequestTimeout = 30 * time.Second
+	// default Timeout to mark Samba service launch as unsuccessful
+	defaultServiceStartTimeOutInSecond = 30
```
nit: slightly shorter name (e.g. _defaultSvcStartTimeoutSec_ or even _defaultStartTimeout_ with comment) would be nice.

msterin commented on docs/user-guide/vfile-plugin.md:
```diff
@@ -3,14 +3,14 @@ title: vFile volume plugin for Docker
 ---
 ## Overview
-Depending on the underlying block storage device system, it might not be possible to access the same 
+Depending on the underlying block storage device system, it might not be possible to access the same
```
`block device`  or `block storage system`, not both

lipingxue replied:
Done.

msterin commented on docs/user-guide/vfile-plugin.md:
````diff
@@ -124,3 +124,11 @@ docker service create --replicas 1 -p 8080:80 --name web nginx
 `Connection refused` error means the routing mesh of this Docker Swarm cluster is broken and needs to be fixed.
+
+### I got " VolumeDriver.Mount: Failed to blocking wait for Mounted state. Error: Timeout reached; BlockingWait is not complete.." when mounting a volume.
+We see this issue only on platforms which the availble disk space are low. On those platforms, docker swarm service takes longer time to up and running. We recommend to run this driver on platforms with enough available disk space. You can also try to increase the service start timeout value by specifing ```VFILE_TIMEOUT_IN_SECOND``` value when installing the plugin:
````
Suggestion: 

When  available disk space are low, Docker Swarm service may take longer start a service. Generally it's better free up some e disk space. You can also try to increase the service start timeout value, controlled by ```VFILE_TIMEOUT_IN_SECOND``` env variable:

lipingxue replied:
Done.

lipingxue replied:
Done.

msterin commented on docs/user-guide/vfile-plugin.md:
````diff
@@ -124,3 +124,11 @@ docker service create --replicas 1 -p 8080:80 --name web nginx
+```
+docker plugin install --grant-all-permissions --alias vfile cnastorage/vfile:latest VFILE_TIMEOUT_IN_SECOND=90
+```
+The above command increases the service start timeout value to 90 seconds and the default value is 30 seconds.
````
This will increase timeout to 90 sec, from default of 30 sec

lipingxue replied:
Done.
